src/PostgREST/Error.hs

```diff
@@ -170,26 +174,78 @@ instance JSON.ToJSON ApiRequestError where
         (True, _, _)                 -> " function with a single json or jsonb parameter"
         (_, True, MTTextPlain)       -> " function with a single unnamed text parameter"
         (_, True, MTTextXML)         -> " function with a single unnamed xml parameter"
         (_, True, MTOctetStream)     -> " function with a single unnamed bytea parameter"
-        (_, True, MTApplicationJSON) -> prms <> " function or the " <> schema <> "." <> procName <>" function with a single unnamed json or jsonb parameter"
-        _                            -> prms <> " function") <>
+        (_, True, MTApplicationJSON) -> fmtParams <> " or the " <> func <>" function with a single unnamed json or jsonb parameter"
+        _                            -> fmtParams) <>
       " in the schema cache"),
     "details" .= JSON.Null,
-    "hint"    .= ("If a new function was created in the database with this name and parameters, try reloading the schema cache." :: Text)]
+    "hint"    .= if hasPreferSingleObject || (isInvPost && contentType `elem` [MTTextPlain, MTTextXML, MTOctetStream])
+                 then Nothing
+                 else noRpcHint schema procName prms allProcs overloadedProcs ]
   toJSON (AmbiguousRpc procs)  = JSON.object [
     "code"    .= SchemaCacheErrorCode03,
     "message" .= ("Could not choose the best candidate function between: " <> T.intercalate ", " [pdSchema p <> "." <> pdName p <> "(" <> T.intercalate ", " [ppName a <> " => " <> ppType a | a <- pdParams p] <> ")" | p <- procs]),
     "details" .= JSON.Null,
     "hint"    .= ("Try renaming the parameters or the function itself in the database so function overloading can be resolved" :: Text)]
 
+-- |
+-- If no function is found with the given name, it does a fuzzy search to all the functions
+-- in the same schema and shows the best match as hint.
+--
+-- >>> :set -Wno-missing-fields
+-- >>> let procsMap = HM.fromList [(QualifiedIdentifier "api" "test" , []), (QualifiedIdentifier "api" "another" , []), (QualifiedIdentifier "private" "other" , [])]
+--
+-- >>> noRpcHint "api" "testt" "(val, param, name)" procsMap []
+-- Just "Perhaps you meant to call the function api.test"
+--
+-- >>> noRpcHint "api" "other" "()" procsMap []
+-- Just "Perhaps you meant to call the function api.another"
+--
+-- >>> noRpcHint "api" "noclosealternative" "()" procsMap []
+-- Nothing
+--
+-- If a function is found with the given name, but no params match, then it does a fuzzy searh
+-- to all the overloaded functions' params using the form "(param1, param2, param3, ...)"
+-- and shows the best match as hint.
+--
+-- >>> let procsDesc = [ProcDescription {pdParams = [ProcParam {ppName="val"}, ProcParam {ppName="param"}, ProcParam {ppName="name"}]}, ProcDescription {pdParams = [ProcParam {ppName="id"}, ProcParam {ppName="attr"}]}]
+--
+-- >>> noRpcHint "api" "test" "(vall, pqaram, nam)" procsMap procsDesc
+-- Just "Perhaps you meant to call the function api.test(val, param, name)"
```

Comment:
This is really cool!

Q: Why do the params use a Text type`"(vall, pqaram, nam)"`? Why can't they be a List?

Reply:
The text type is needed because fuzzy search looks for a Text in a set of possible matches, that's why the solution I saw was formatting the parameters as Text. So, behind the scenes it does this:

`Fuzzy.getOne (Fuzzy.fromList ["(val, param, name)", "(id, attr)"])  "(vall, pqaram, nam)"`

And returns the first one directly to be used by the hint. Although, the arguments could be formatted into a Text inside the function itself, but I wanted to reuse the `prms` function that already did that.

Reply:
Hm, I still don't understand. From the [fuzzy set docs](https://hackage.haskell.org/package/fuzzyset-0.2.3/docs/Data-FuzzySet.html) I don't see that they use that format.
 
```haskell
Fuzzy.getOne (Fuzzy.fromList ["(val, param, name)", "(id, attr)"]) "(vall, pqaram, nam)"
```

So the fuzzy search is ignoring the parentheses?

Reply:
Yes, it's doing that. I separated that login in other function and added a comment to hopefully make it clearer.